### ISSUE TITLE
Rewrite of cleanup block fails when using multiple assignment of method return value

### ISSUE THREAD

**Issue by luisgoncalves** [spockframework/spock#1413]
### Describe the bug

A test method with:
- a `cleanup` block
- assignment of multiple variables from the return of an *instance* method

Causes compilation to fail. If the method used in the multiple assignment is made static, compilation succeeds.

### To Reproduce

```
class ExampleSpec extends Specification {

    def "Tuple with cleanup"() {
        when:
        def (first, second) = getArray()

        then:
        first
        !second

        cleanup:
        println 'cleanup'
    }

    private Boolean[] getArray() {
        [true, false]
    }
}
```

### Expected behavior

Specification compiles.

### Actual behavior

Compilation fails with:

```
Unexpected error during compilation of spec 'ExampleSpec'. Maybe you have used invalid Spock syntax? Anyway, please file a bug report at http://issues.spockframework.org.
java.lang.NullPointerException
	at org.spockframework.compiler.SpecRewriter.transformRhsExpressionIfNecessary(SpecRewriter.java:777)
	at org.spockframework.compiler.SpecRewriter.moveVariableDeclarations(SpecRewriter.java:763)
	at org.spockframework.compiler.SpecRewriter.visitCleanupBlock(SpecRewriter.java:459)
	at org.spockframework.compiler.model.CleanupBlock.accept(CleanupBlock.java:33)
	at org.spockframework.compiler.model.Method.accept(Method.java:70)
	at org.spockframework.compiler.model.Spec.accept(Spec.java:112)
	at org.spockframework.compiler.SpockTransform$Impl.processSpec(SpockTransform.java:76)
	at org.spockframework.compiler.SpockTransform$Impl.visit(SpockTransform.java:63)
	at org.spockframework.compiler.SpockTransform.visit(SpockTransform.java:48)
	at org.codehaus.groovy.transform.ASTTransformationVisitor.lambda$addPhaseOperationsForGlobalTransforms$4(ASTTransformationVisitor.java:337)
	at org.codehaus.groovy.control.CompilationUnit$ISourceUnitOperation.doPhaseOperation(CompilationUnit.java:901)
	at org.codehaus.groovy.control.CompilationUnit.processPhaseOperations(CompilationUnit.java:671)
	at org.codehaus.groovy.control.CompilationUnit.compile(CompilationUnit.java:635)
	at org.codehaus.groovy.control.CompilationUnit.compile(CompilationUnit.java:610)
	at org.gradle.api.internal.tasks.compile.ApiGroovyCompiler.execute(ApiGroovyCompiler.java:262)
	at org.gradle.api.internal.tasks.compile.ApiGroovyCompiler.execute(ApiGroovyCompiler.java:64)
	at org.gradle.api.internal.tasks.compile.GroovyCompilerFactory$DaemonSideCompiler.execute(GroovyCompilerFactory.java:97)
	at org.gradle.api.internal.tasks.compile.GroovyCompilerFactory$DaemonSideCompiler.execute(GroovyCompilerFactory.java:76)
	at org.gradle.api.internal.tasks.compile.daemon.AbstractDaemonCompiler$CompilerWorkAction.execute(AbstractDaemonCompiler.java:135)
	at org.gradle.workers.internal.DefaultWorkerServer.execute(DefaultWorkerServer.java:63)
	at org.gradle.workers.internal.AbstractClassLoaderWorker$1.create(AbstractClassLoaderWorker.java:49)
	at org.gradle.workers.internal.AbstractClassLoaderWorker$1.create(AbstractClassLoaderWorker.java:43)
	at org.gradle.internal.classloader.ClassLoaderUtils.executeInClassloader(ClassLoaderUtils.java:97)
	at org.gradle.workers.internal.AbstractClassLoaderWorker.executeInClassLoader(AbstractClassLoaderWorker.java:43)
	at org.gradle.workers.internal.IsolatedClassloaderWorker.run(IsolatedClassloaderWorker.java:49)
	at org.gradle.workers.internal.IsolatedClassloaderWorker.run(IsolatedClassloaderWorker.java:30)
	at org.gradle.workers.internal.WorkerDaemonServer.run(WorkerDaemonServer.java:85)
	at org.gradle.workers.internal.WorkerDaemonServer.run(WorkerDaemonServer.java:55)
	at org.gradle.process.internal.worker.request.WorkerAction$1.call(WorkerAction.java:138)
	at org.gradle.process.internal.worker.child.WorkerLogEventListener.withWorkerLoggingProtocol(WorkerLogEventListener.java:41)
	at org.gradle.process.internal.worker.request.WorkerAction.run(WorkerAction.java:135)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:566)
	at org.gradle.internal.dispatch.ReflectionDispatch.dispatch(ReflectionDispatch.java:36)
	at org.gradle.internal.dispatch.ReflectionDispatch.dispatch(ReflectionDispatch.java:24)
	at org.gradle.internal.remote.internal.hub.MessageHubBackedObjectConnection$DispatchWrapper.dispatch(MessageHubBackedObjectConnection.java:182)
	at org.gradle.internal.remote.internal.hub.MessageHubBackedObjectConnection$DispatchWrapper.dispatch(MessageHubBackedObjectConnection.java:164)
	at org.gradle.internal.remote.internal.hub.MessageHub$Handler.run(MessageHub.java:414)
	at org.gradle.internal.concurrent.ExecutorPolicy$CatchAndRecordFailures.onExecute(ExecutorPolicy.java:64)
	at org.gradle.internal.concurrent.ManagedExecutorImpl$1.run(ManagedExecutorImpl.java:48)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628)
	at org.gradle.internal.concurrent.ThreadFactoryImpl$ManagedThreadRunnable.run(ThreadFactoryImpl.java:56)
	at java.base/java.lang.Thread.run(Thread.java:834)
```

If `getArray` is made static, compilation succeeds.

### Java version

```
openjdk 11.0.2 2019-01-15
OpenJDK Runtime Environment 18.9 (build 11.0.2+9)
OpenJDK 64-Bit Server VM 18.9 (build 11.0.2+9, mixed mode)
```

### Buildtool version

```
------------------------------------------------------------
Gradle 7.1
------------------------------------------------------------

Build time:   2021-06-14 14:47:26 UTC
Revision:     989ccc9952b140ee6ab88870e8a12f1b2998369e

Kotlin:       1.4.31
Groovy:       3.0.7
Ant:          Apache Ant(TM) version 1.10.9 compiled on September 27 2020
JVM:          11.0.2 (Oracle Corporation 11.0.2+9)
OS:           Mac OS X 10.16 x86_64

```

### What operating system are you using

Mac

### Dependencies

```
dependencies {
    implementation localGroovy()
    testImplementation "org.spockframework:spock-core:2.0-groovy-3.0"
}
```

**Comment by luisgoncalves** [comment]
The following compiles:

```
def x = getArray()
def (first, second) = x
```

**Comment by kriegaex** [comment]
Note to myself and anyone else who is not a Groovy pro: We are talking about [this nice Groovy feature](https://blog.mrhaki.com/2009/09/groovy-goodness-multiple-assignments.html).

**Comment by kriegaex** [comment]
Hm, I tried this with Groovy 3.0.7 and 3.0.9, Spock 2.0 and 2.1-M2. The test always passes. I am using Maven, however, and the Eclipse Batch Compiler. I am too lazy to try and reproduce your exact situation, but chances are, that this is not a Spock issue. Can you publish an [MCVE](https://stackoverflow.com/help/mcve) on GitHub, maybe?

**Comment by renatoathaydes** [comment]
You can run this class with plain Groovy like this:

* add a `@Grab` declaration so groovy itself sets up the classpath:

```groovy
@Grab('org.spockframework:spock-core:2.0-groovy-3.0')
import spock.lang.Specification

class ExampleSpec extends Specification {

    def "Tuple with cleanup"() {
        when:
        def (first, second) = getArray()

        then:
        first
        !second

        cleanup:
        println 'cleanup'
    }

    private Boolean[] getArray() {
        [true, false]
    }
}

// print something as when running with Groovy, a script is expected
println 'Compilation successful'
```

Run with `groovy your-file.groovy`.

Here's what I get:

```
org.codehaus.groovy.control.MultipleCompilationErrorsException: startup failed:
scratch.groovy: Unexpected error during compilation of spec 'ExampleSpec'. Maybe you have used invalid Spock syntax? Anyway, please file a bug report at http://issues.spockframework.org.

java.lang.NullPointerException
        at org.spockframework.compiler.SpecRewriter.transformRhsExpressionIfNecessary(SpecRewriter.java:777)
        at org.spockframework.compiler.SpecRewriter.moveVariableDeclarations(SpecRewriter.java:763)
        at org.spockframework.compiler.SpecRewriter.visitCleanupBlock(SpecRewriter.java:459)
        at org.spockframework.compiler.model.CleanupBlock.accept(CleanupBlock.java:33)
        at org.spockframework.compiler.model.Method.accept(Method.java:70)
        at org.spockframework.compiler.model.Spec.accept(Spec.java:112)
        at org.spockframework.compiler.SpockTransform$Impl.processSpec(SpockTransform.java:76)
        at org.spockframework.compiler.SpockTransform$Impl.visit(SpockTransform.java:63)
        at org.spockframework.compiler.SpockTransform.visit(SpockTransform.java:48)
        at org.codehaus.groovy.transform.ASTTransformationVisitor.lambda$addPhaseOperationsForGlobalTransforms$4(ASTTransformationVisitor.java:337)
        at org.codehaus.groovy.control.CompilationUnit$ISourceUnitOperation.doPhaseOperation(CompilationUnit.java:901)
        at org.codehaus.groovy.control.CompilationUnit.processPhaseOperations(CompilationUnit.java:671)
        at org.codehaus.groovy.control.CompilationUnit.compile(CompilationUnit.java:635)
        at groovy.lang.GroovyClassLoader.doParseClass(GroovyClassLoader.java:389)
        at groovy.lang.GroovyClassLoader.lambda$parseClass$3(GroovyClassLoader.java:332)
        at org.codehaus.groovy.runtime.memoize.StampedCommonCache.compute(StampedCommonCache.java:163)
        at org.codehaus.groovy.runtime.memoize.StampedCommonCache.getAndPut(StampedCommonCache.java:154)
        at groovy.lang.GroovyClassLoader.parseClass(GroovyClassLoader.java:330)
        at groovy.lang.GroovyShell.parseClass(GroovyShell.java:526)
        at groovy.lang.GroovyShell.run(GroovyShell.java:359)
        at groovy.lang.GroovyShell.run(GroovyShell.java:349)
        at groovy.ui.GroovyMain.processOnce(GroovyMain.java:652)
        at groovy.ui.GroovyMain.run(GroovyMain.java:398)
        at groovy.ui.GroovyMain.access$1400(GroovyMain.java:68)
        at groovy.ui.GroovyMain$GroovyCommand.process(GroovyMain.java:322)
        at groovy.ui.GroovyMain.processArgs(GroovyMain.java:142)
        at groovy.ui.GroovyMain.main(GroovyMain.java:115)
        at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
        at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
        at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
        at java.base/java.lang.reflect.Method.invoke(Method.java:566)
        at org.codehaus.groovy.tools.GroovyStarter.rootLoader(GroovyStarter.java:111)
        at org.codehaus.groovy.tools.GroovyStarter.main(GroovyStarter.java:129)


1 error
```

Tested with Groovy versions:

* Groovy Version: 3.0.6 JVM: 11.0.3 Vendor: Azul Systems, Inc. OS: Mac OS X
* Groovy Version: 3.0.9 JVM: 11.0.3 Vendor: Azul Systems, Inc. OS: Mac OS X
* Groovy Version: 3.0.9 JVM: 17 Vendor: Oracle Corporation OS: Mac OS X

All of these result in the same error.

Using the workaround @luisgoncalves showed, it compiles fine.

So it does look like it's unrelated to Gradle, very surprising Maven would be able to compile this somehow.

**Comment by renatoathaydes** [comment]
Oh, maybe you confused your Spock version, because I've just tested with the upcoming Spock 2.1 and it does seem to work with that.

Change the Grab annotation to:

```
@Grab('org.spockframework:spock-core:2.1-M2-groovy-3.0')
```

Then the problem seems to have been fixed in spock 2.1. I think we can afford to upgrade to this milestone... will let you know if we find any problems.

BTW this syntax can even use type declarations:

```groovy
def (boolean first, boolean second)  = getArray()
```

We use this in many places :)

**Comment by kriegaex** [comment]
Hmm... Why did it work with Spock 2.0 for me too? Currently I am not near a computer, but maybe neither my IDE nor Maven recompiled the spec for some reason. I might have forgotten `clean`. Going to try again later.

**Comment by kriegaex** [comment]
> Why did it work with Spock 2.0 for me too? (...) I might have forgotten `clean`.

That was it, yes. I now can confirm that in my project the test also fails to compile in 2.0 and works since 2.1-M1. So it looks as if this issue is really resolved. But a Spock committer should check if it was fixed intentionally or could still be broken in other edge cases.

**Comment by leonard84** [comment]
This is a duplicate of #1332 and is already fixed in 2.1-M2
